src/frontends/sql/decoder_context.cc

```diff
@@ -22,33 +23,13 @@ const Operation &DecoderContext::MakeMergeOperation(
 
 const ir::Operation &DecoderContext::MakeTagTransformOperation(
     ir::Value transformed_value, absl::string_view rule_name,
-    const absl::flat_hash_map<std::string, uint64_t> &preconditions) {
-  // We need to reserve one space for the transformed value plus one for each
-  // precondition.
-  ir::ValueList inputs;
-
-  // Insert the value to be transformed.
-  inputs.push_back(transformed_value);
-
-  // Each precondition is given the same input name prefix plus an incrementing
-  // number in the same order as in the original ID list.
-  ir::NamedAttributeMap attributes;
-
-  attributes.insert({std::string(kTagTransformRuleAttributeName),
-                     ir::Attribute::Create<ir::StringAttribute>(rule_name)});
-  int64_t index = 1;
-  for (const auto &[name, id] : preconditions) {
-    CHECK(index != std::numeric_limits<int64_t>::max())
-        << "`int64_t` might overflow. Why would you create so many attributes?";
-    inputs.push_back(GetValue(id));
-    auto insert_result = attributes.insert(
-        {name, ir::Attribute::Create<ir::Int64Attribute>(index++)});
-    CHECK(insert_result.second)
-        << "Found duplicate entry for " << insert_result.first->first;
+    const absl::flat_hash_map<std::string, uint64_t> &precondition_indices) {
+  std::vector<std::pair<std::string, ir::Value>> preconditions;
+  for (const auto &[name, id] : precondition_indices) {
+    preconditions.push_back({name, GetValue(id)});
   }
-
-  return top_level_block_builder_.AddOperation(tag_transform_operator_,
-                                               attributes, inputs, {});
+  return top_level_block_builder_.AddOperation<TagTransformOp>(
+      ir_context_, rule_name, transformed_value, preconditions);
```

Comment:
Is this a good place for `MapIter`?

Reply:
I just tried it out, but the current snippet is more readable. MapIter is quite verbose due to (1) lack of destructuring in the lambda, (2) explicit call to `std::make_pair`, etc.